File: design-pattern-java/app/src/main/java/design/pattern/java/patterns/structural/proxy/ServiceCacheFactory.java
package design.pattern.java.patterns.structural.proxy;

import java.util.HashMap;
import java.util.Map;

public class ServiceCacheFactory {
  private static ServiceCacheFactory instance;

  public static ServiceCacheFactory create() {
    if (instance == null) {
      ServiceCacheFactory.instance = new ServiceCacheFactory();
    }
    return ServiceCacheFactory.instance;
  }

  private final Map<String, Object> cache;

  private ServiceCacheFactory() {
    this.cache = new HashMap<>();
  }

  public Boolean isCached(final String paramsString) {
    return this.cache.containsKey(paramsString);
  }

  public <T> T saveInCache(String paramString, T value) {
    this.cache.put(paramString, value);
    return value;
  }

  public Object getInCache(String paramString) {
    return this.cache.get(paramString);
  }
}

File: design-pattern-jvm/app/src/main/kotlin/design/pattern/kotlin/patterns/structural/proxy/ServiceCacheFactory.kt
package design.pattern.kotlin.patterns.structural.proxy

import java.util.HashMap

class ServiceCacheFactory private constructor() {
    private val cache: MutableMap<String, Any>
    fun isCached(paramsString: String): Boolean {
        return cache.containsKey(paramsString)
    }

    fun <T : Any> saveInCache(paramString: String, value: T): T {
        cache[paramString] = value
        return value
    }

    fun getInCache(paramString: String): Any? {
        return cache[paramString]
    }

    companion object {
        private var instance: ServiceCacheFactory? = null
        fun create(): ServiceCacheFactory? {
            if (instance == null) {
                instance = ServiceCacheFactory()
            }
            return instance
        }
    }

    init {
        cache = HashMap()
    }
}